library/std/src/sys/dbg.rs

```diff
@@ -0,0 +1,216 @@
+            Ok(DebuggerPresence::NotDetected)
+        }
+    }
+}
+
+#[cfg(any(target_os = "macos", target_os = "freebsd"))]
+mod os {
+    use libc::{c_int, sysctl, CTL_KERN, KERN_PROC, KERN_PROC_PID};
+
+    use super::{DebuggerPresence, DebuggerPresenceError};
+    use crate::io::{Read, Seek};
+    use crate::process;
+
+    const P_TRACED: i32 = 0x00000800;
+
+    // The assumption is that the kernel structures available to the
+    // user space may not shrink or repurpose the existing fields over
+    // time. The kernels normally adhere to that for the backward
+    // compatibility of the user space.
+
+    // macOS 14.6.1 uses a 3,888 byte structure; using the smaller size
+    // to support older systems. Validated by compiling a C program
+    // without reliance on a specific OS version.
+    #[cfg(all(target_os = "macos", target_pointer_width = "64"))]
+    const KINFO_PROC_SIZE: usize = 648;
+    #[cfg(all(target_os = "macos", target_pointer_width = "64"))]
+    const KINFO_PROC_FLAGS_OFFSET: u64 = 32;
+
+    // 32-bit macOS Rust targets are no longer shipped.
+    // Make it so the logic below returns "NotImpelemnted".
+    #[cfg(all(target_os = "macos", target_pointer_width = "32"))]
+    const KINFO_PROC_SIZE: usize = 0;
+    #[cfg(all(target_os = "macos", target_pointer_width = "32"))]
+    const KINFO_PROC_FLAGS_OFFSET: u64 = 0;
+
+    // Works for FreeBSD stable (13.3, 13.4) and current (14.0, 14.1).
+    // The size of the structure has stayed the same for a long time,
+    // at list since 2005:
+    // https://lists.freebsd.org/pipermail/freebsd-stable/2005-November/019899.html
+    #[cfg(all(target_os = "freebsd", target_pointer_width = "64"))]
+    const KINFO_PROC_SIZE: usize = 1088;
+    #[cfg(all(target_os = "freebsd", target_pointer_width = "64"))]
+    const KINFO_PROC_FLAGS_OFFSET: u64 = 368;
+
+    #[cfg(all(target_os = "freebsd", target_pointer_width = "32"))]
+    const KINFO_PROC_SIZE: usize = 768;
+    #[cfg(all(target_os = "freebsd", target_pointer_width = "32"))]
+    const KINFO_PROC_FLAGS_OFFSET: u64 = 296;
+
+    pub(super) fn is_debugger_present() -> Result<DebuggerPresence, DebuggerPresenceError> {
+        if KINFO_PROC_SIZE == 0 {
+            return Err(DebuggerPresenceError::NotImplemented);
+        }
+
+        let mut flags = [0u8; 4]; // `ki_flag` under FreeBSD and `p_flag` under macOS.
+        let mut mib = [CTL_KERN, KERN_PROC, KERN_PROC_PID, process::id() as c_int];
+        let mut info_size = 0;
+
+        // SAFETY: No state is shared with other threads. The sysctl call
+        // is safe according to the documentation.
+        // This particular form of the call with NULL pointers is used to
+        // read the size of the `kinfo_proc` structure into `info_size`.
+        if unsafe {
+            sysctl(
+                mib.as_mut_ptr(),
+                mib.len() as u32,
+                core::ptr::null_mut(),
+                &mut info_size,
+                core::ptr::null_mut(),
```

Comment:
nit: Why do you not also `use core::ptr;` in this module?

Reply:
Bad manners on my side 😸 Fixed, thanks!